### PR TITLE
Add support for mounting volumes with local driver and options

### PR DESCRIPTION
**Pull request by mheon**
Volumes with the `local` driver and mount options specified are mounted using the `mount` command, instead of being plain directories. The volumes are mounted as the first container using them is started, and unmounted when the last container using them exits.

Given the mount/unmount behavior, we need to add a mount counter to volumes to know when to mount/unmount. This requires adding a mutable state to volumes, with associated methods in the DB to update and save the state.

Missing tests right now, but I've tested locally with tmpfs and btrfs with positive results.

### CODE REVIEW

**Comment by mheon** [comment]
@rhatdan PTAL - you said this was blocking your work on overlay volumes

**Comment by TomSweeneyRedHat** [comment]
I think not, but thought I'd ask.  Any anticipated db upgrade issues?

**Comment by mheon** [comment]
The first commit should address most of them.

The bigger issue was the locks PR that went in earlier - we'll need a `podman system renumber` if any volumes were present when the upgrade happened.

**Comment by TomSweeneyRedHat** [comment]
Thanks @mheon.  A quick browse and the code LGTM.  The `podman system renumber` bit I'm assuming will be loud and proud in the appropriate release announcement and out on the Podman mailing list for starters.

File: libpod/boltdb_state.go [hunk 1]
```diff
@@ -1352,6 +1400,16 @@ func (s *BoltState) AddVolume(volume *Volume) error {
...
 	}
 
+	// Volume state is allowed to not exist
+	var volStateJSON []byte
+	if volume.state != nil {
+		stateJSON, err := json.Marshal(volume.state)
+		if err != nil {
+			return errors.Wrapf(err, "error marshalling volume %s state to JSON", volume.Name())
+		}
+		volStateJSON = stateJSON
```
Comment (baude) [inline libpod/boltdb_state.go]: any reason not to volStateJSON, err = json.Marshal?

**Comment by baude** [comment]
lgtm other than nit question ... have you measured any perf impact to overall container running with this?

**Comment by mheon** [comment]
Should be 0 on containers that do not mount shared volumes.

On those with volumes requiring a full mount call, on first mount, it's pretty substantial - around 1s extra for a tmpfs, in my measurements. This about matches what I saw with Docker, though - was how I figured out they had to be doing the mount on container launch.

File: libpod/boltdb_state.go (hunk 1 above)
Comment (mheon) [reply libpod/boltdb_state.go]: I think `json.Marshal` on a nil pointer makes angry errors

Comment (mheon) [reply libpod/boltdb_state.go]: Oh wait, i see what you're saying. Probably will work.

File: libpod/container_internal.go [hunk 2]
```diff
@@ -1225,6 +1225,40 @@ func (c *Container) mountStorage() (string, error) {
...
+				vol.lock.Unlock()
+				return "", errors.Wrapf(err, "error mounting volume %s for container %s", vol.Name(), c.ID())
+			}
+			vol.lock.Unlock()
+			defer func() {
+				if Err == nil {
+					return
+				}
+				vol.lock.Lock()
+				defer vol.lock.Unlock()
```
Comment (vrothberg) [inline libpod/container_internal.go]: If we move it down a few lines, we can spare the `defer` and save a few cycles.

File: libpod/runtime_volume_linux.go [hunk 3]
```diff
@@ -48,6 +48,19 @@ func (r *Runtime) newVolume(ctx context.Context, options ...VolumeCreateOption)
 	}
 	volume.config.CreatedTime = time.Now()
 
+	if volume.config.Driver == "local" {
```
Comment (vrothberg) [inline libpod/runtime_volume_linux.go]: Same here: can we use a constant for "local"?

File: libpod/volume_internal_linux.go [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+	}
+	if volType != "" {
+		mountArgs = append(mountArgs, "-t", volType)
+	}
+	mountArgs = append(mountArgs, volDevice, v.config.MountPoint)
+	mountCmd := exec.Command(mountPath, mountArgs...)
+
+	errPipe, err := mountCmd.StderrPipe()
+	if err != nil {
+		return errors.Wrapf(err, "getting stderr pipe")
```
Comment (vrothberg) [inline libpod/volume_internal_linux.go]: "error getting ..."?

File: libpod/volume_internal_linux.go [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+	// give the kernel something.
+	if volDevice == "" && volType != "" {
+		volDevice = volType
+	}
+
+	// We need to use the actual mount command.
+	// Convincing unix.Mount to use the same semantics as the mount command
+	// itself seems prohibitively difficult.
+	// TODO: might want to cache this path in the runtime?
+	mountPath, err := exec.LookPath("mount")
```
Comment (vrothberg) [inline libpod/volume_internal_linux.go]: I'd prefer calling the syscall directly. I am no mount expert, so I may be totally unaware of the difficulties.

File: pkg/varlinkapi/volumes.go [hunk 6]
```diff
@@ -67,7 +68,8 @@ func (i *LibpodAPI) GetVolumes(call iopodman.VarlinkCall, args []string, all boo
 			Labels:     v.Labels(),
 			MountPoint: v.MountPoint(),
 			Name:       v.Name(),
-			Options:    v.Options(),
+			// TODO change types here to be correct
+			//Options:    v.Options(),
```
Comment (vrothberg) [inline pkg/varlinkapi/volumes.go]: Don't we need the options as well?

File: libpod/volume_internal_linux.go (hunk 5 above)
Comment (mheon) [reply libpod/volume_internal_linux.go]: My initial attempt tried that, but I abandoned it pretty quickly due to difficulties syncing up the defaults. Docker allows semi-arbitrary options to be passed, does almost no authentication, and pops them straight into a call to `mount`. We would need to reverse-engineer somewhat large tracts of the functionality of Coreutils `mount` to make this work properly without it.

File: pkg/varlinkapi/volumes.go (hunk 6 above)
Comment (mheon) [reply pkg/varlinkapi/volumes.go]: Oh, oops, that one's a mistake, reverting

**Comment by rhatdan** [comment]
I would like to see @vrothberg questions addressed, as well as squash the PRs, and add some tests.  Examples in man pages would also be nice.

Looks very good.

File: libpod/volume.go [hunk 7]
```diff
@@ -26,15 +29,15 @@ type VolumeConfig struct {
...
+	Driver string `json:"volumeDriver"`
 	// The location the volume is mounted at.
 	MountPoint string `json:"mountPoint"`
 	// Time the volume was created.
 	CreatedTime time.Time `json:"createdAt,omitempty"`
 	// Options to pass to the volume driver. For the local driver, this is
 	// a list of mount options. For other drivers, they are passed to the
 	// volume driver handling the volume.
-	Options map[string]string `json:"options"`
+	Options map[string]string `json:"volumeOptions,omitempty"`
```
[the remark this answers is not in the document]
Comment (mheon) [reply libpod/volume.go]: We need to discard anything already in the database. Previously we were doing absolutely no validation on these fields - now we're properly vetting their contents. We can't go back and make previous containers safe, so the next best option is to ignore them (they were all acting like volumes without options prior to this PR anyways)

**Comment by mheon** [comment]
Addressed review comments

**Comment by rhatdan** [comment]
@mheon can you squash your commits.

**Comment by mheon** [comment]
Sure

**Comment by mheon** [comment]
Added tests

**Comment by mheon** [comment]
Assuming things go green, I think we're good for final review and merge

**Comment by mheon** [comment]
Going green. @rhatdan @baude @vrothberg @TomSweeneyRedHat @giuseppe PTAL

**Pull request merged**